services/middleware/src/transactions.rs

```diff
@@ -0,0 +1,192 @@
+        artifact_name: "test-artifact".to_string(),
+        artifact_version: "^1.0.0".parse().unwrap(),
+        inner: AnyTx {
+            call_info: CallInfo::new(100, 0),
+            arguments: vec![],
+        },
+    };
+    assert_eq!(
+        serde_json::to_value(&checked_call).unwrap(),
+        json!({
+            "artifact_name": "test-artifact",
+            "artifact_version": "^1.0.0",
+            "inner": checked_call.inner,
+        })
+    );
+
+    checked_call.artifact_version = ">=0.9, <2".parse().unwrap();
+    assert_eq!(
+        serde_json::to_value(&checked_call).unwrap(),
+        json!({
+            "artifact_name": "test-artifact",
+            "artifact_version": ">= 0.9, < 2",
+            "inner": checked_call.inner,
+        })
+    );
+}
+
+/// Transactions executed in a batch.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[derive(ProtobufConvert, BinaryValue)]
+#[protobuf_convert(source = "proto::Batch")]
+pub struct Batch {
+    /// Transactions included in the batch.
+    pub inner: Vec<AnyTx>,
+}
+
+impl Batch {
+    /// Creates an empty batch.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Appends a call into the batch.
+    pub fn with_call(mut self, call: AnyTx) -> Self {
+        self.inner.push(call);
+        self
+    }
+}
+
+/// Transactional interface of the utilities service.
+#[exonum_interface]
+pub trait UtilsInterface {
+    /// Performs a checked call to the service. The call is dispatched only if the version
+    /// of the service matches the version requirement mentioned in the call.
+    ///
+    /// # Authorization
+    ///
+    /// The inner call is authorized in the same way as the `checked_call`.
+    fn checked_call(
+        &self,
+        context: CallContext<'_>,
+        arg: CheckedCall,
+    ) -> Result<(), ExecutionError>;
+
+    /// Performs batch execution of several transactions. Transactions are executed
+    /// in the order they are mentioned in the batch. If execution of the constituent transaction
+    /// fails, the method returns an error, thus rolling back any changes performed by earlier
+    /// transactions.
+    ///
+    /// # Authorization
+    ///
+    /// All transactions are authorized in the same way as the `batch` call itself.
+    fn batch(&self, context: CallContext<'_>, arg: Batch) -> Result<(), ExecutionError>;
+}
+
+impl UtilsInterface for MiddlewareService {
+    fn checked_call(
+        &self,
+        mut context: CallContext<'_>,
+        arg: CheckedCall,
+    ) -> Result<(), ExecutionError> {
+        let instance_id = arg.inner.call_info.instance_id;
+        let dispatcher_schema = context.data().for_dispatcher();
+        let state = dispatcher_schema
+            .get_instance(instance_id)
+            .ok_or(Error::NoService)?;
+        if state.status != InstanceStatus::Active {
+            return Err(Error::ServiceIsNotActive.into());
+        }
+
+        let artifact = &state.spec.artifact;
```

Comment:
Is this approach assume that the service interfaces completely change with every change in the version of the artifact?

Reply:
The caller may specify the version requirement in the form he is comfortable with. For example, he may specify the version requirement as `^1.1.3`, in which case any semver-compatible version will match (e.g., `1.1.5`, `1.3.0` or `1.5.9`). If the caller is brave enough, he may specify a more wide range of versions, e.g., `>=1.1.3, <3`. Or he may be more conservative and specify `~1.1.3` (which means `>=1.1.3, <1.2.0`).

What is expected from the service developers is that artifact versioning reflects the public interfaces of the service (i.e., at least transactions), so that semver requirements make sense.